presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduSplitManager.java

```diff
@@ -59,4 +81,60 @@ public ConnectorSplitSource getSplits(
                 throw new IllegalArgumentException("Unknown splitSchedulingStrategy: " + splitSchedulingStrategy);
         }
     }
+
+    private static CompletableFuture<?> whenCompleted(DynamicFilter dynamicFilter)
+    {
+        if (dynamicFilter.isAwaitable()) {
+            return dynamicFilter.isBlocked().thenCompose(ignored -> whenCompleted(dynamicFilter));
+        }
+        else {
+            return NOT_BLOCKED;
+        }
+    }
+
+    private static class KuduDynamicFilteringSplitSource
+            implements ConnectorSplitSource
+    {
+        private final CompletableFuture<?> dynamicFilterFuture;
+        private final CompletableFuture<ConnectorSplitSource> splitSourceFuture;
+
+        private KuduDynamicFilteringSplitSource(
+                CompletableFuture<?> dynamicFilterFuture,
+                CompletableFuture<ConnectorSplitSource> splitSourceFuture)
+        {
+            this.dynamicFilterFuture = requireNonNull(dynamicFilterFuture, "dynamicFilterFuture is null");
+            this.splitSourceFuture = requireNonNull(splitSourceFuture, "splitSourceFuture is null");
+        }
+
+        @Override
+        public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
+        {
+            return splitSourceFuture.thenCompose(splitSource -> splitSource.getNextBatch(partitionHandle, maxSize));
+        }
+
+        @Override
+        public void close()
+        {
+            if (!dynamicFilterFuture.cancel(true)) {
```

Comment:
Could it be that cancellation succeeded and `splitSourceFuture` created split source anyway (I think it still might be the case).

What if we implement `getNextBatch` as:
```
@GuardedBy("this")
private Optional<ConnectorSplitSource> delegateSplitSource = Optional.empty();

public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
{
  if (!dynamicFilterFuture.isDone()) {
    return dynamicFilterFuture.thenApply(new ConnectorSplitSource.ConnectorSplitBatch(ImmutableList.of(), false));
  }

  return getDelegateSplitSource().getNextBatch(partitionHandle, maxSize)
}

private synchronized ConnectorSplitSource getDelegateSplitSource() {
  if (delegateSplitSource.isPresent()) {
   return delegateSplitSource.get();
  }

  delegateSplitSource = Optional.of(...)
  return delegateSplitSource.get();
}
```

Reply:
> Could it be that cancellation succeeded and splitSourceFuture created split source anyway (I think it still might be the case).
I do not think the case exist.
If dynamicFilterFuture cancellation succeeded, then dynamicFilterFuture is not completed normally, then function getSplitSource() is not triggered.
Reasoning backward, if getSplitSource() is already triggered, we know dynamicFilterFuture is completed normally, dynamicFilterFuture cancellation should fail.

Reply:
>Could it be that cancellation succeeded and splitSourceFuture created split source anyway (I think it still might be the case).

I do not think the case should exist.
If dynamicFilterFuture cancellation succeeded, then dynamicFilterFuture is not completed normally, then function getSplitSource() is not triggered.
Reasoning backward, if getSplitSource() is already triggered, we know dynamicFilterFuture is completed normally, dynamicFilterFuture cancellation should fail.